aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -82,69 +77,76 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
 Tensor& randperm_out_cuda(Tensor& result, int64_t n, c10::optional<Generator> generator) {
   TORCH_CHECK(n >= 0, "n must be non-negative, got", n);
   TORCH_CHECK(!generator.has_value() || (generator.has_value() && result.device() == generator->device()), "Expected a '", result.device(), "' generator device but found '", generator->device(), "'");
+  TORCH_CHECK(n <= std::numeric_limits<int>::max(),
+    "randperm of tensors larger than INT_MAX is not supported yet in pytorch");
   check_supported_max_int_with_precision(n, result);
 
   result.resize_({n});
 
-  if (n < 30000) {  // For small inputs, we offload it to CPU instead.
-    auto result_cpu = at::empty({n}, result.options().device(kCPU));
-    randperm_out(result_cpu, n, generator);
-    return result.copy_(result_cpu);
+  auto range = at::arange(n, result.options());
+
+  // shuffled_data points to the underlying data of the output tensor if the tensor is contiguous; otherwise it
+  // points to a new tensor.
+  Tensor shuffled;
+  void *shuffled_data;
+  if (result.is_contiguous()) {
+    shuffled_data = result.data_ptr();
+  } else {
+    shuffled = at::empty(n, result.options());
+    shuffled_data = shuffled.data_ptr();
   }
 
-#if 0
-  // This if condition should never be true because if n >= 30000 and the tensor has a Half type,
-  // check_supported_max_int_with_precision should have reported an error. This snippet is commented out but left here
-  // for the sake of clarity, because Half in thrust is spotty, and we do not want future change unaware of this.
-  if (result.scalar_type() == at::ScalarType::Half) {  // Half in thrust is spotty. Avoid.
-    auto result_float = at::empty({n}, initialTensorOptions().device(Device(DeviceType::CUDA)));
-    return result.copy_(randperm_out_cuda(result_float, n, generator));
+  auto opt = TensorOptions().device(result.device());
+
+  int bits = std::min(64,
+    static_cast<int>(std::ceil(std::log2(static_cast<double>(n)))) + 1);
```

Comment:
do you need +1 if you already have `ceil`? E.g. if `n=4` you need to sort on 2 bits?

Reply:
See https://github.com/pytorch/pytorch/pull/54113#discussion_r596535150, I am just adding one extra mantissa to reduce the chance of two keys being the same, which will make the result more random.